tests/ztimer_periodic/main.c

```diff
@@ -32,62 +32,79 @@
 
 static mutex_t _mutex = MUTEX_INIT_LOCKED;
 
-#define MAX_OFFSET 2
-#define N 5
-#define INTERVAL 100LU
+#define N 6
 
 static uint32_t _times[N];
+static int _count;
+#define CLOCKS { ZTIMER_MSEC, ZTIMER_USEC }
+static const char *_names[] = { "ZTIMER_MSEC", "ZTIMER_USEC" };
+static uint32_t _intervals[] = { 100, 10000 };
+static uint32_t _max_offsets[] = { 2, 100 };
 
 static int callback(void *arg)
 {
-    (void)arg;
-    static int count = 0;
+    if (_count == 0) {
+        _count++;
+        return 0;
+    }
 
-    _times[count] = ztimer_now(ZTIMER_MSEC);
+    _times[_count] = ztimer_now(arg);
 
-    count += 1;
+    _count += 1;
 
     /* enable this to test underflow behavior */
 #if 0
     if (count == 2) {
-        ztimer_spin(ZTIMER_MSEC, INTERVAL*2);
+        ztimer_spin(arg, INTERVAL * 2);
     }
 #endif
 
-    if (count == N) {
+    if (_count == N) {
         mutex_unlock(&_mutex);
     }
 
-    return (count == N);
+    return (_count == N);
 }
 
 int main(void)
 {
     ztimer_periodic_t t;
+    ztimer_clock_t * const clocks[] = CLOCKS;
+    int failed = 0;
 
-    ztimer_periodic_init(ZTIMER_MSEC, &t, callback, NULL, INTERVAL);
-    uint32_t last = ztimer_now(ZTIMER_MSEC);
-
-    ztimer_periodic_start(&t);
-
-    if (!ztimer_is_set(ZTIMER_MSEC, &t.timer)) {
-        print_str("Test failed\n");
-        return 1;
-    }
+    for (size_t j = 0; j < ARRAY_SIZE(clocks); j++) {
+        printf("Testing clock: %s\n", _names[j]);
+        ztimer_clock_t *clock = clocks[j];
+        ztimer_periodic_init(clock, &t, callback, clock, _intervals[j]);
+        uint32_t last = ztimer_now(clock);
```

Comment:
The other option is to use the internal value of `ztimer_periodic_t`:

```suggestion
        uint32_t last = t.last;
```

Reply:
Done

Reply:
Re-running the tests

Reply:
The situation improved a bit (offset is smaller) but the same BOARD's are still failing... :(

### arduino-uno/failuresummary

Failures during test:
- [tests/ztimer_periodic](https://ci.inria.fr/ci-riot-tribe/job/build-pipeline.jk/267/artifact/atmega256rfr2-xpro/tests/ztimer_periodic/test.failed/*view*/)

### z1/failuresummary

Failures during test:
- [tests/ztimer_periodic](https://ci.inria.fr/ci-riot-tribe/job/build-pipeline.jk/267/artifact/z1/tests/ztimer_periodic/test.failed/*view*/)

### atmega256rfr2-xpro/failuresummary

Failures during test:
- [tests/ztimer_periodic](https://ci.inria.fr/ci-riot-tribe/job/build-pipeline.jk/267/artifact/arduino-uno/tests/ztimer_periodic/test.failed/*view*/)

Reply:
On the arduino-uno, the `ZTIMER_USEC` times are consistently bad, while on atmega256rfr2-xpro it works fine for all but the first timeout. I think both are running at 16 MHz, are using the same peripheral time and should have (almost) the same IPC, so I wonder what is causing the difference.

Reply:
So how do we move forward?

Reply:
I'd say the test is fine, but the implementation in AVR is flanky. I would like to debug this, but I currently don't have the time :-/